src/util/ConfigManager/ConfigManager.h

```diff
@@ -0,0 +1,247 @@
+  entire path in json, needed to address the value held by this option.
+  @param optionDescription A description for the configuration option.
+  @param variableToPutValueOfTheOptionIn The value held by the configuration
+  option will be copied into this variable, whenever the value in the
+  configuration option changes.
+  @param defaultValue A default value for the configuration option. If none is
+  given, signified by an empty optional, then a value for the configuration
+  option MUST be given at runtime.
+  */
+  template <typename OptionType>
+  requires ad_utility::isTypeContainedIn<OptionType,
+                                         ConfigOption::AvailableTypes>
+  void createConfigOption(std::string_view optionName,
+                          std::string_view optionDescription,
+                          OptionType* variableToPutValueOfTheOptionIn,
+                          std::optional<OptionType> defaultValue =
+                              std::optional<OptionType>(std::nullopt)) {
+    createConfigOption<OptionType>(
+        VectorOfKeysForJson{std::string{optionName}}, optionDescription,
+        variableToPutValueOfTheOptionIn, defaultValue);
+  }
+
+  /*
+  @brief Get all the added configuration options.
+  */
+  const std::vector<ConfigOption>& getConfigurationOptions() const;
+
+  /*
+  @brief Sets the configuration options based on the given json. Note: This will
+  overwrite values held by the configuration data, if there were values for them
+  given inside the string.
+
+  @param j There will be an exception thrown, if:
+  - `j` doesn't contain values for all configuration options, that must be set
+  at runtime.
+
+  - Same, if there are values for configuration options, that do not exist.
+
+  - `j` is anything but a json object literal.
+  */
+  void parseConfig(const nlohmann::json& j);
+
+  /*
+  @brief Parses the given short hand and returns it as a json object,
+   that contains all the described configuration data.
+
+  @param shortHandString The language of the short hand is defined in
+  `generated/ConfigShorthand.g4`.
+  */
+  static nlohmann::json parseShortHand(const std::string& shortHandString);
+
+  /*
+  @brief Returns a string containing a json configuration and the
+  string representations of all added configuration options.
+
+  @param printCurrentJsonConfiguration If true, the current values of the
+  configuration options will be used for printing the json configuration. If
+  false, an example json configuration will be printed, that uses the default
+  values of the configuration options, and, if an option doesn't have a default
+  value, a dummy value.
+  */
+  std::string printConfigurationDoc(bool printCurrentJsonConfiguration) const;
+
+  /*
+  @brief Return a string, containing a list of all configuration options, that:
+  - Have a default value.
+  - Weren't set at runtime.
+  */
+  std::string getListOfNotChangedConfigOptionsWithDefaultValues() const;
```

Comment:
I didn't do it exactly like that, but the function now returns a vector and I wrote a `AsString` function. See [30d203b](https://github.com/ad-freiburg/qlever/pull/977/commits/30d203ba90aa66259ba518a7042ffe8b79220525).

Would that be alright?

Reply:
After reading the rest of the PR I now have the following realization:

Maybe you don't need this function to be public at all (The `BenchmarkMain` should use the `printConfigurationDoc` anyway). And then you can leave this function as a stringj returning function, because it is just an internal helper function to format errors. What do you think about this?

Reply:
> Maybe you don't need this function to be public at all (The `BenchmarkMain` should use the `printConfigurationDoc` anyway). And then you can leave this function as a stringj returning function, because it is just an internal helper function to format errors. What do you think about this?

Do I understand correctly, that you want to make `getListOfNotChangedConfigOptionsWithDefaultValues` private and only `getListOfNotChangedConfigOptionsWithDefaultValuesAsString` public?

P.S.: And what do do you mean with

> (The `BenchmarkMain` should use the printConfigurationDoc anyway)

? It already does, but only when you use `-o` to show all available configuration options. (I didn't want to clutter up the output by printing the huge list everytime.)

Reply:
@ PS:
I meant that the `"print"` case currently prints some partial configurations using these functions. I think you should just print the complete configuration there (it is easy to scroll down) using the `printConfigurationDoc` and THEN you need none of the two functions to be public. And then I think if you only need the string version private, you can only keep that one (accessing the options from inside the class is much easier). But I still think, none of this is needed outside.

Reply:
Not 100% sure, I understood, what you wanted, but I made an attempt in [0c94109](https://github.com/ad-freiburg/qlever/pull/977/commits/0c9410912a5ab27a1588c6c66d83059aa87ac93d).